numpy/lib/recfunctions.py

```diff
@@ -870,6 +870,330 @@ def repack_fields(a, align=False, recurse=False):
     dt = np.dtype(fieldinfo, align=align)
     return np.dtype((a.type, dt))
 
+def _get_fields_and_offsets(dt, offset=0):
+    """
+    Returns a flat list of (name, dtype, count, offset) tuples of all the
+    scalar fields in the dtype "dt", including nested fields, in left
+    to right order.
+    """
+    fields = []
+    for name in dt.names:
+        field = dt.fields[name]
+        if field[0].names is None:
+            count = 1
+            for size in field[0].shape:
+                count *= size
+            fields.append((name, field[0], count, field[1] + offset))
+        else:
+            fields.extend(_get_fields_and_offsets(field[0], field[1] + offset))
+    return fields
+
+
+def _structured_to_unstructured_dispatcher(arr, dtype=None, copy=None,
+                                           casting=None):
+    return (arr,)
+
+@array_function_dispatch(_structured_to_unstructured_dispatcher)
+def structured_to_unstructured(arr, dtype=None, copy=False, casting='unsafe'):
```

Comment:
I think there are places we learnt that `unsafe` was a bad default, but ended up stuck with it, leaving users surprised by the conversion.

Should we apply that learning here, and pick a more conservative default?

Reply:
I may have missed that. Maybe in #8733? But that was about assignment using `unsafe` casting, with no option to specify otherwise, unlike here where there is a keyword the user can specify.

Reply:
Yeah, that issue was one of the ones I was thinking of - thanks for linking it, I was looking for that for unrelated reasons too!

Unsafe just feels like an... unsafe default to me. In my opinion, unsafe behavior should be something you ask for, not something you get by default. You're picking between:

* as is: `f(...)` vs `f(..., casting='safe')`
* proposed: `f(..., casting='unsafe')` vs `f(...)`

I'd much rather see the word 'unsafe' to tell me I need to think more carefully about that line of code, rather than having to look for the absence of it.

I don't have a good memory of how the other casting modes behave. I'd be inclined to pick `same_kind` to match the default value of the casting argument for for `ufunc`s

Reply:
Here's an argument for `unsafe`: 

First, it matches the default for the same keyword in `astype`, and so its easier for the user to remember if they are used to using `astype`. 

Second, it seems like most of the time the user *wants* `unsafe`, because there are many common casts that are ruled out otherwise. For instance casts from `f8` to `i8` are disallowed with `same_kind`, but I expect this is a very common cast.

Actually, for reasons I don't understand, ufuncs seem to allow casts from `f8` to `i8` even though they supposedly use `same_kind`:
```python
>>> np.arange(3, dtype='f8').astype('i8', casting='same_kind')                 
TypeError: Cannot cast array from dtype('float64') to dtype('int64') according to the rule 'same_kind'
>>> np.add(np.arange(3, dtype='f8'), np.arange(3, dtype='i8'))                 
array([0., 2., 4.])
```
So ufuncs appear to use `unsafe` casting despite the keyword default??